arch/x86/chorba_sse41.c

```diff
@@ -0,0 +1,344 @@
+        c = _mm_xor_si128(_mm_srli_epi64(invec, 45), _mm_slli_epi64(invec, 44)); \
+        d  = _mm_srli_epi64(invec, 20); \
+        } while (0);
+
+#define REALIGN_CHORBA(in0, in1, in2, in3, out0, out1, out2, out3, out4, shift) do { \
+        out0 = _mm_slli_si128(in0, shift); \
+        out1 = _mm_alignr_epi8(in1, in0, shift); \
+        out2 = _mm_alignr_epi8(in2, in1, shift); \
+        out3 = _mm_alignr_epi8(in3, in2, shift); \
+        out4 = _mm_srli_si128(in3, shift); \
+        } while (0)
+
+#define STORE4(out0, out1, out2, out3, out) do { \
+        _mm_store_si128(out++, out0); \
+        _mm_store_si128(out++, out1); \
+        _mm_store_si128(out++, out2); \
+        _mm_store_si128(out++, out3); \
+    } while (0)
+
+#define READ4(out0, out1, out2, out3, in) do { \
+    out0 = _mm_load_si128(in++); \
+    out1 = _mm_load_si128(in++); \
+    out2 = _mm_load_si128(in++); \
+    out3 = _mm_load_si128(in++); \
+    } while (0)
+
+/* This is intentionally shifted one down to compensate for the deferred store from
+ * the last iteration */
+#define READ4_WITHXOR(out0, out1, out2, out3, xor0, xor1, xor2, xor3, in) do { \
+    out0 = _mm_xor_si128(_mm_load_si128(&in[1]), xor0); \
+    out1 = _mm_xor_si128(_mm_load_si128(&in[2]), xor1); \
+    out2 = _mm_xor_si128(_mm_load_si128(&in[3]), xor2); \
+    out3 = _mm_xor_si128(_mm_load_si128(&in[4]), xor3); \
+    } while (0)
+
+static Z_FORCEINLINE uint32_t crc32_chorba_32768_nondestructive_sse41(uint32_t crc, const uint64_t* buf, size_t len) {
+    const uint64_t* input = buf;
+    ALIGNED_(16) uint64_t bitbuffer[32768 / sizeof(uint64_t)];
+    const uint8_t* bitbufferbytes = (const uint8_t*) bitbuffer;
+    __m128i z = _mm_setzero_si128();
+
+    __m128i *bitbuf128 = (__m128i*)(bitbuffer + 128);
+    __m128i *bitbuf144 = (__m128i*)(bitbuffer + 144);
+    __m128i *bitbuf182 = (__m128i*)(bitbuffer + 182);
+    __m128i *bitbuf0 = bitbuf128;
+
+    /* We only need to zero out the bytes between the 128'th value and the 144th
+     * that are actually read */
+    for (size_t i = 0; i < 2; ++i) {
+        STORE4(z, z, z, z, bitbuf128);
+    }
+
+    /* We only need to zero out the bytes between the 144'th value and the 182nd that
+     * are actually read */
+    for (size_t i = 0; i < 11; ++i) {
+        _mm_store_si128(&bitbuf144[8], z);
+        ++bitbuf144;
+    }
+
+    /* We only need to zero out the bytes between the 182nd value and the 210th that
+     * are actually read. */
+    for (size_t i = 0; i < 4; ++i) {
+        STORE4(z, z, z, z, bitbuf182);
+    }
+
+    /* We need to mix this in */
+    __m128i init_crc = _mm_cvtsi64x_si128(crc);
+    crc = 0;
+
+    size_t i = 0;
+    __m128i *inptr = (__m128i*)input;
+    bitbuf144 = (__m128i*)(bitbuffer + 144);
+    bitbuf182 = (__m128i*)(bitbuffer + 182);
+    __m128i *bitbuf210 = (__m128i*)(bitbuffer + 210);
+    __m128i *bitbuf300 = (__m128i*)(bitbuffer + 300);
```

Comment:
Should use temporary variable to avoid excessive casting, array notation as above...